lighthouse.go

```diff
@@ -53,6 +55,114 @@ func NewLightHouse(amLighthouse bool, myIp uint32, ips []uint32, interval int, n
 	return &h
 }
 
+func NewLightHouseFromConfig(config *Config, myIp uint32, nebulaPort int, pc *udpConn) (*LightHouse, error) {
+	h := LightHouse{
+		amLighthouse: config.GetBool("lighthouse.am_lighthouse", false),
+		myIp:         myIp,
+		addrMap:      make(map[uint32][]udpAddr),
+		nebulaPort:   nebulaPort,
+		lighthouses:  make(map[uint32]struct{}),
+		staticList:   make(map[uint32]struct{}),
+		interval:     config.GetInt("lighthouse.interval", 10),
+		punchConn:    pc,
+		punchBack:    config.GetBool("punch_back", false),
+	}
+
+	err := h.loadStaticHosts(config)
+	if err != nil {
+		return nil, err
+	}
+
+	config.RegisterReloadCallback(h.reload)
+
+	return &h, nil
+}
+
+func (lh *LightHouse) loadStaticHosts(config *Config) error {
+	rawLighthouseHosts := config.GetStringSlice("lighthouse.hosts", []string{})
+	if lh.amLighthouse && len(rawLighthouseHosts) != 0 {
+		l.Warn("lighthouse.am_lighthouse enabled on node but upstream lighthouses exist in config")
+	}
+
+	for i, host := range rawLighthouseHosts {
+		ip := net.ParseIP(host)
+		if ip == nil {
+			l.WithField("host", host).Fatalf("Unable to parse lighthouse host entry %v", i+1)
+		}
+		lh.lighthouses[ip2int(ip)] = struct{}{}
+	}
+
+	for k, v := range config.GetMap("static_host_map", map[interface{}]interface{}{}) {
+		vpnIp := net.ParseIP(fmt.Sprintf("%v", k))
+		vals, ok := v.([]interface{})
+		if ok {
+			for _, v := range vals {
+				err := lh.addStaticRemote(vpnIp, v)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			err := lh.addStaticRemote(vpnIp, v)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	err := lh.ValidateLHStaticEntries()
+	if err != nil {
+		l.WithError(err).Error("Lighthouse unreachable")
+	}
+
+	return nil
+}
+
+func (lh *LightHouse) reload(config *Config) {
+	if config.GetBool("lighthouse.am_lighthouse", false) != lh.amLighthouse {
+		l.Warn("Changing lighthouse.am_lighthouse with SIGHUP not supported. Ignored.")
+	}
+	lh.interval = config.GetInt("lighthouse.interval", 10)
+	lh.punchBack = config.GetBool("punch_back", false)
+
+	err := lh.loadStaticHosts(config)
+	if err != nil {
+		l.WithError(err).Error("failed to reload static hosts")
+	}
+
+	// Check for removed lighthouses / static entries. load the new config
+	// into a fresh LightHouse and then see which entries are no longer there.
+
+	newLh := LightHouse{
+		addrMap:     make(map[uint32][]udpAddr),
+		lighthouses: make(map[uint32]struct{}),
+		staticList:  make(map[uint32]struct{}),
+	}
+	newLh.loadStaticHosts(config)
+
+	var removedStatic, removedLighthouses []uint32
+	for vpnIP := range lh.staticList {
+		if _, ok := newLh.staticList[vpnIP]; !ok {
+			removedStatic = append(removedStatic, vpnIP)
+		}
+	}
+	for vpnIP := range lh.lighthouses {
+		if _, ok := newLh.lighthouses[vpnIP]; !ok {
+			removedLighthouses = append(removedLighthouses, vpnIP)
+		}
+	}
+
+	for _, vpnIP := range removedLighthouses {
```

Comment:
This seems highly likely to cause a concurrent read/write access panic in both maps and our current RWMutex is entirely for the `addrMap`, seems we need at least one more now.

This also appears to only remove the diff set and not add any new hosts?